W-20001897-add-CH2DRlink-410-dm (#3236)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/hadr-guide.adoc

== High Availability Versus Disaster Recovery

> * *High availability (HA)* - The measure of a system's ability to remain accessible in the event of a system component failure. Generally, HA is implemented by building in multiple levels of fault tolerance and load balancing capabilities into a system.

Reviewer: ```suggestion
* *High availability (HA)* - The measure of a system's ability to remain accessible in the event of a system component failure. Generally, HA is implemented by building multiple levels of fault tolerance and load balancing capabilities into a system.
```

* *Disaster recovery (DR)* - The process by which a system is restored to a previous acceptable state, after a natural (flooding, tornadoes, earthquakes, fires) or man-made (power failures, server failures, misconfigurations) disaster.

While they both increase overall availability, the notable difference is that with HA there is generally no loss of service. HA retains the service and DR retains the data, but with DR, there is usually a slight loss of service while the DR plan executes and the system restores.

== Deploying Mule for HA and DR Strategies

> You can deploy the Mule runtime in many different topologies to address your HA and DR strategies. One method is through the use of clustering, which is described later in this document. Mule High Availability Clustering provides basic failover capability for Mule.

Reviewer: ```suggestion
You can deploy the Mule runtime in many different topologies to address your HA and DR strategies. One method is through the use of high availability clustering, which provides basic failover capability for Mule.
```

When the primary Mule runtime becomes unavailable, for example, because of a fatal JVM or hardware failure or it's taken offline for maintenance, a backup Mule runtime immediately becomes the primary node and resumes processing where the failed instance left off.

After a system administrator recovers a failed Mule runtime server and puts it back online, that server automatically becomes the backup node.

image::hadr-cluster.png["A clustered server environment handling incoming and outgoing messages"]

Seamless failover is made possible by a distributed memory store that shares all transient state information among clustered Mule runtimes, such as:

* SEDA service event queues
* In-memory message queues

As you build your Mule application, it's important to think critically about how best to architect your application to achieve the desired availability, fault tolerance, and performance characteristics.

== Creating Effective SLAs

The Service Level Agreement (SLA) outlines the specific definition of what is acceptable.

The basic definition of high availability is that the service is up and

> functioning, that is, properly processing requests and responses. This doesn't

Reviewer: ```suggestion
functioning and properly processing requests and responses. This doesn't
```

*Costs*

> Vary depending on SLA requirements. This model needs 2 nodes to satisfy SLA. If the SLA's acceptable service impact changes to the terms stated in the normal operations then at a minimum the environment needs 3 nodes to accommodate 1 node failure. More nodes may be required depending on the probability of not having at least 2 nodes running.

Reviewer: ```suggestion
Vary depending on SLA requirements. This model needs two nodes to satisfy SLA. If the SLA's acceptable service impact changes to the terms stated in the normal operations then at a minimum the environment needs three nodes to accommodate one node failure. More nodes may be required depending on the probability of not having at least two nodes running.
```

=== Active-Active Clustering Fault Tolerance Deployment Model

> The basic definition of fault tolerance is a failure within the system doesn't impact the service at all. This differs from high availability as service impact and downtime is tolerated.

Reviewer: ```suggestion
Fault tolerance means a failure within the system doesn't impact the service at all. This differs from high availability as service impact and downtime are tolerated.
```

=== Zero Downtime Deployment Model

The goal is to make changes to the environment without impacting the SLAs; including upgrading infrastructure and the applications running on the infrastructure. Typically zero downtime deployments leverage a side-by-side deployment, where the old and new coexist for a short period of time. This is in contrast to an in-place deployment where the service may experience reduced capacity to complete downtime.

http://www.gartner.com/it-glossary/continuous-operations/[Gartner] defines continuous operations as “those characteristics of a data-processing system that reduce or eliminate the need for planned downtime, such as scheduled maintenance. One element of 24-hour-a-day, seven-day-a-week operation”.

> The baseline production environment is the current operating environment. A new environment is created with the changes (upgraded runtimes, configurations, new applications).  A small percentage of traffic flows to the new environment and increases as the confidence in the new environment increases. The baseline production environment continues its use until the new environment is fully operational (it's handling 100% of the traffic). Once the new environment is accepting all traffic, it becomes the new baseline production environment and the previous baseline production environment terminates.

Reviewer: ```suggestion
The baseline production environment is the current operating environment. A new environment is created with the changes (upgraded runtimes, configurations, new applications).  A small percentage of traffic flows to the new environment and increases as the confidence in the new environment increases. The baseline production environment continues its use until the new environment is fully operational (it's handling 100% of the traffic). After the new environment is accepting all traffic, it becomes the new baseline production environment and the previous baseline production environment terminates.
```

> This example is assuming each environment is using the same number of Mule runtimes and cores. It's plausible that the new environment may have more or less number than the baseline environment.

Reviewer: ```suggestion
This example is assuming each environment is using the same number of Mule runtimes and cores. It's plausible that the new environment has either more or fewer Mule runtimes and cores than the baseline environment.
```

*How quickly can your company get back to work after an IT emergency?*

> Disaster recovery (DR) is the process by which a system is restored to a previous acceptable state, after a natural or man-made disaster. It's important that the selection of the appropriate DR strategy be driven by business requirements. For DR, use measurable characteristics, such as Recovery Time Objective (RTO) and Recovery Point Objective (RPO), to drive your DR plan.

Reviewer: ```suggestion
Disaster recovery (DR) is the process by which a system is restored to a previous acceptable state, after a natural or man-made disaster. For DR, use measurable business requirements such as Recovery Time Objective (RTO) and Recovery Point Objective (RPO), to drive your DR plan.
```

> Disaster recovery is about your Recovery Point Objective (RPO) and your Recovery Time Objective (RTO). RPO is the "point" that you return to after an IT disaster. For example, if you back up every 24 hours, your RPO is a maximum of 24 hours ago. RTO, on the other hand, is how quickly you can restore to your RPO and get back to business. This includes activities like the time it takes to get your spare equipment to start running your backups if your primary equipment isn't working.

Reviewer: ```suggestion
Disaster recovery is about your Recovery Point Objective (RPO) and your Recovery Time Objective (RTO). RPO is the "point" that you return to after an IT disaster. For example, if you back up the system every 24 hours, your RPO is a maximum of the past 24 hours. RTO, on the other hand, is how quickly you can restore to your RPO and get back to business. This includes activities like the time it takes to get your spare equipment to start running your backups if your primary equipment isn't working.
```

|Zero data loss, 15 mins out of date?
|===

For deployment-specific disaster recovery details, see:

* xref:runtime-manager::cloudhub-hadr.adoc[].
* xref:cloudhub-2::ch2-ha-dr.adoc[].

== Keep Integrations Stateless

> As a general design principle it's always important to ensure the integrations are stateless in nature. This means that no transactional information is shared between various client invocations or the executions (in case of scheduled services). If some data has to be maintained by the middleware due to a system limitation, it should be persisted in an external store such as a database or a messaging queue and not within the middleware infrastructure or memory. It's critical to note that as we scale, especially in the cloud, the state and resources used by each worker/node should be independent of the other worker. This model ensures better performance, scalability as well as reliability.

Reviewer: ```suggestion
As a general design principle it's important to ensure integrations are stateless in nature. This means that no transactional information is shared between various client invocations or the executions (in case of scheduled services). If the middleware needs to keep some data because of a system limitation, it should be persisted in an external store such as a database or a messaging queue. It shouldn't be stored within the middleware infrastructure or memory. It's critical to note that as the system scales, especially in the cloud, the state and resources used by each worker/node should be independent of the other workers. This model ensures better performance, scalability, and reliability.
```
